File: docs/dashboard/src/palettes.ts
/**
 * Client-side mirror of tools/palettes.py — the dashboard's PMTiles rasters are
 * pre-colorized server-side (build_dashboard_assets.py), so there is no palette
 * metadata to read from a data file at runtime. These constants must stay
 * byte-for-byte in sync with tools/palettes.py so the map legends (MapLegend.vue)
 * describe the colors actually baked into each raster.
 */

export const PAL_SUIT = ['#d7191c', '#fdae61', '#ffffbf', '#a6d96a', '#1a9641'] // 0..1 red->green
export const PAL_FAO = ['#d7191c', '#fdae61', '#a6d96a', '#1a9641'] // N, S3, S2, S1
export const PAL_ROLE = ['#eeeeee', '#ffd400', '#7b3294', '#d95f0e', '#2c7fb8', '#addd8e', '#006837']
export const PAL_DIV = ['#b2182b', '#f7f7f7', '#2166ac'] // diverging deltaS
export const PAL_AGREEMENT = ['#ffffcc', '#a1dab4', '#41b6c4', '#225ea8'] // sequential, GCM sign-agreement
export const PAL_UNDERUSED = ['#f7f7f7', '#d7301f'] // binary, potential-vs-realized gap

export const FAO_ORDER = ['N', 'S3', 'S2', 'S1'] as const

/** rl_role band codes (src/external.py ROLE_CODES), 0 = other/none, in PAL_ROLE order. */
export const ROLE_LABELS_PT = [
  'Outro / nenhum',
  'Soja',
  'Cana-de-açúcar',
  'Outras culturas anuais',
  'Piscicultura',
  'Pastagem',
  'Vegetação nativa',
]

const TAB10 = [
  '#1f77b4', '#ff7f0e', '#2ca02c', '#d62728', '#9467bd',
  '#8c564b', '#e377c2', '#7f7f7f', '#bcbd22', '#17becf',
]
const TAB20 = [
  '#1f77b4', '#aec7e8', '#ff7f0e', '#ffbb78', '#2ca02c', '#98df8a', '#d62728', '#ff9896',
  '#9467bd', '#c5b0d5', '#8c564b', '#c49c94', '#e377c2', '#f7b6d2', '#7f7f7f', '#c7c7c7',
  '#bcbd22', '#dbdb8d', '#17becf', '#9edae5',
]

/**
 * Mirrors tools/palettes.py's zone_palette(n): samples matplotlib's discrete
 * tab10/tab20 ListedColormap the same way (floor(i * N / (n-1)), clipped to N-1),
 * so a given zone count always maps to the exact same colors as the server-side
 * `zones_present.pmtiles` raster.
 */
export function zonePalette(n: number): string[] {
  const table = n <= 10 ? TAB10 : TAB20
  const size = table.length
  if (n <= 1) return [table[0]!]
  return Array.from({ length: n }, (_, i) => {
    const idx = Math.min(size - 1, Math.floor((i * size) / (n - 1)))
    return table[idx]!
  })
}

File: tools/palettes.py
"""Shared vis-param/color-ramp definitions.

Single source of truth for palettes used by both `tools/make_figures.py` (thesis PNGs)
and `tools/build_dashboard_assets.py` (dashboard PMTiles/Plotly), so the two never drift
into different color scales for the same layer (docs/dashboard_ux_plan.md §2.2).

These hex lists are also mirrored by hand into `docs/dashboard/src/styles/tokens.css`.
"""
from __future__ import annotations

from matplotlib import colormaps
from matplotlib.colors import rgb2hex

PAL_SUIT = ["#d7191c", "#fdae61", "#ffffbf", "#a6d96a", "#1a9641"]  # 0..1 red->green
PAL_FAO = ["#d7191c", "#fdae61", "#a6d96a", "#1a9641"]              # N, S3, S2, S1
PAL_ZONE = ["#4575b4", "#91bfdb", "#f46d43", "#fdae61", "#66c2a5", "#1a9850", "#762a83"]  # 7-zone legacy constant (kept for make_figures.py backward-compat; see zone_palette() for the current zone count)
PAL_ROLE = ["#eeeeee", "#ffd400", "#7b3294", "#d95f0e", "#2c7fb8", "#addd8e", "#006837"]
PAL_DIV = ["#b2182b", "#f7f7f7", "#2166ac"]                          # diverging ΔS
PAL_PR = ["#f7fbff", "#9ecae1", "#3182bd", "#08306b"]                 # sequential, precipitation
PAL_TEMP = ["#fff5eb", "#fd8d3c", "#d94801", "#7f2704"]               # sequential, temperature
PAL_AGREEMENT = ["#ffffcc", "#a1dab4", "#41b6c4", "#225ea8"]          # sequential, GCM sign-agreement
PAL_UNDERUSED = ["#f7f7f7", "#d7301f"]                                # binary, potential-vs-realized gap

SEGMENTS = [
    "soybean",
    "sugarcane",
    "other_crops",
    "pisciculture",
    "cattle",
    "conservation",
    "solar",
]

FAO_CODES = {"N": 0, "S3": 1, "S2": 2, "S1": 3}
FAO_ORDER = ["N", "S3", "S2", "S1"]


def zone_palette(n: int) -> list[str]:
    """A categorical palette with exactly `n` colors, for whatever the current zone
    count actually is (zoning is re-run under different K over time — see
    zone_profiles.csv row count for ground truth, not a hardcoded literal)."""
    cmap = colormaps["tab10"] if n <= 10 else colormaps["tab20"]
    return [rgb2hex(cmap(i / max(n - 1, 1))) for i in range(n)]
